### PR TITLE
feat: Adding guidance on commit message style

### LINKED ISSUE #2050: guidance on commit message style
Would be great to add some guidance on what your commit message should look like when contributing a PR to [kubernetes/kubernetes](github.com/kubernetes/kubernetes) if there is a format/convention. Some repos have a specific commit message style that helps with generating release notes. See [this one](http://docs.deis.io/en/latest/contributing/standards/#commit-style) as an example.

### PR DESCRIPTION
**Pull request by vishakhanihore**
**Which issue(s) this PR fixes**:

Fixes #2050 

- This PR concerns with adding a section for `Git Commit Message`.

### CODE REVIEW

**Comment by vishakhanihore** [comment]
cc: @nikhita @mrbobbytables @markyjackson-taulia

**Comment by vishakhanihore** [comment]
cc: @markyjackson-taulia for `lgtm`

**Comment by castrojo** [comment]
/ok-to-test
/hold

When it comes to something like commit styles and messages, I'd like to get consensus from the larger contributor community, perhaps post on the contributor-experience mailing list pointing to this issue so we can feedback.

File: contributors/guide/pull-requests.md [hunk 1]
```diff
@@ -280,26 +277,41 @@ Each commit should have a good title line (<70 characters) and include an additi
...
  Don't squash when there are independent changes layered to achieve a single goal. For instance, writing a code munger could be one commit, applying it could be another, and adding a precommit check could be a third. One could argue they should be separate pull requests, but there's really no way to test/review the munger without seeing it applied, and there needs to be a precommit check to ensure the munged output doesn't immediately get out of date.
 
-A commit, as much as possible, should be a single logical change.
-
-## 7. KISS, YAGNI, MVP, etc.
+## 7. Commit Message Guidelines
+ 
+- Each commit should have a good title line (< 50 characters) and can include an additional description paragraph in the PR Description describing the change in more detail.
+- Wrap the body of the commit message at 72 characters and use the body to explain **What and Why** is the change and **How** it has been done.
+- It is better if the Pull Request follow the `Semantic PR Commit Guidelines` which is `{Type}: {Subject}`. **Types** can be classified as :
```
Comment (saschagrunert) [inline contributors/guide/pull-requests.md]: Would it generally make sense to stick to `Type == kind`? If yes, we could link the kind label to this sort of typing guideline.

Comment (mrbobbytables) [reply contributors/guide/pull-requests.md]: I think that's a great idea. Even if there is more than one kind applied to it, can defer to what it primarily is.

Comment (vishakhanihore) [reply contributors/guide/pull-requests.md]: >  If yes, we could link the kind label to this sort of typing guideline.

Sounds great! I don't think currently I have the right to create labels right now.

Comment (mrbobbytables) [reply contributors/guide/pull-requests.md]: We wouldn't create new labels, but use the ones we currently use. For reference, here is our PR template for the k/k repo:
https://raw.githubusercontent.com/kubernetes/kubernetes/master/.github/PULL_REQUEST_TEMPLATE.md

In there we have a block:
```

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind deprecation
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake
```

That would add some context to the commit without having to reference the labels associated with the PR.

Comment (tpepper) [reply contributors/guide/pull-requests.md]: I think it's also reasonable to use "{area}" in addition to kind.

File: contributors/guide/pull-requests.md [hunk 2]
```diff
@@ -280,26 +277,41 @@ Each commit should have a good title line (<70 characters) and include an additi
...
+    + `feat` : A PR concerning with a new feature.
+    + `fix` : A PR concerning with a bug fix.
+    + `improvement` : A PR concerning with an improvement to a current feature.
+    + `docs` : A PR which only make changes to the documentation.
+    + `refactor` : A PR concerning with a code change that neither fixes a bug nor adds a feature.
+    + `test` : A PR which adds missing tests or corrects the existing tests.
+    + `chore` : Other changes which don't modify src or test files and can be regarded as maintenance check-up for the same.
+- Additional Readings :
+    + [Git Commit](https://chris.beams.io/posts/git-commit/)
```
Comment (tpepper) [inline contributors/guide/pull-requests.md]: Yes!  This does a good job of describing the  "what and why" distinction from "how".

File: contributors/guide/pull-requests.md (hunk 1 above)
Comment (mrbobbytables) [reply contributors/guide/pull-requests.md]: Agree 👍

**Comment by mrbobbytables** [comment]
The commit message guideline section has been reworked if folks wouldn't mind taking a look again 👍 

/hold

File: contributors/guide/pull-requests.md [hunk 3]
```diff
@@ -25,11 +25,22 @@ This doc explains the process and best practices for submitting a pull request t
...
+    - [The first word in the commit message subject should be capitalized unless it starts with a lowercase symbol or other identifier](#the-first-word-in-the-commit-message-subject-should-be-capitalized-unless-it-starts-with-a-lowercase-symbol-or-other-identifier)
+    - [Do not end the commit message subject with a period](#do-not-end-the-commit-message-subject-with-a-period)
+    - [Use imperative mood in your commit message subject](#use-imperative-mood-in-your-commit-message-subject)
+    - [Add a single blank line before the commit message body](#add-a-single-blank-line-before-the-commit-message-body)
+    - [Wrap the commit message body at 72 characters](#wrap-the-commit-message-body-at-72-characters)
+    - [Do not use GitHub keywords or (@)mentions within your commit message](#do-not-use-github-keywords-or-mentions-within-your-commit-message)
+      - [GitHub Keywords](#github-keywords)
+      - [(@)Mentions](#mentions)
+    - [Use the commit message body to explain the _what_ and _why_ of the commit](#use-the-commit-message-body-to-explain-the-what-and-why-of-the-commit)
```
Comment (mrbobbytables) [inline contributors/guide/pull-requests.md:39]: These were added by auto-toc - I'm not quite sure how I feel about them being in here. What are other folks opinions?

Comment (tpepper) [reply contributors/guide/pull-requests.md:39]: Strange...Why would the bullet list interpret as headings?

File: contributors/guide/pull-requests.md [hunk 4]
```diff
@@ -302,34 +313,228 @@ For more information, see [squash commits](./github-workflow.md#squash-commits).
...
+messages are the _"permanent record"_ of the changes being done in your PR and
+their commit messages should accurately describe both _what_ and _why_ it is
+being done.
 
-Sometimes we need to remind each other of core tenets of software design - Keep It Simple, You Aren't Gonna Need It, Minimum Viable Product, and so on. Adding a feature "because we might need it later" is antithetical to software that ships. Add the things you need NOW and (ideally) leave room for things you might need later - but don't implement them now.
+Commit messages are comprised of two parts; the subject and the body.
 
-## 8. It's OK to Push Back
+The subject is the first line of the commit message and is often the only part
+of the commit message needed for small changes. These are often done as "one
```
Comment (tpepper) [inline contributors/guide/pull-requests.md:326]: Could this be more clear?  Eg:

...of the commit message needed for the smallest of trivial changes.  Those may be done as "one liners" with the git `-m` or `--message` flag, but only if the _what_ and especially _why_ can be fully described in that few words.  Typing a few further sentences of clarification is a useful investment in time both for your reviews and overall later project maintenance.

File: contributors/guide/pull-requests.md [hunk 5]
```diff
@@ -302,34 +313,228 @@ For more information, see [squash commits](./github-workflow.md#squash-commits).
...
+
+- [Try to keep the subject line to 50 characters or less; do not exceed 72 characters](#try-to-keep-the-subject-line-to-50-characters-or-less-do-not-exceed-72-characters)
+- [The first word in the commit message subject should be capitalized unless it starts with a lowercase symbol or other identifier](#the-first-word-in-the-commit-message-subject-should-be-capitalized-unless-it-starts-with-a-lowercase-symbol-or-other-identifier)
+- [Do not end the commit message subject with a period](#do-not-end-the-commit-message-subject-with-a-period)
+- [Use imperative mood in your commit message subject](#use-imperative-mood-in-your-commit-message-subject)
+- [Add a single blank line before the commit message body](#add-a-single-blank-line-before-the-commit-message-body)
+- [Wrap the commit message body at 72 characters](#wrap-the-commit-message-body-at-72-characters)
+- [Do not use GitHub keywords or (@)mentions within your commit message](#do-not-use-github-keywords-or-mentions-within-your-commit-message)
+- [Use the commit message body to explain the _what_ and _why_ of the commit](#use-the-commit-message-body-to-explain-the-what-and-why-of-the-commit)
+
```
Comment (tpepper) [inline contributors/guide/pull-requests.md:362]: Could this list be split or ordered by importance, eg: first content needs (what, why, concise summary), second formatting needs (blank space after subject line), and formatting desires (first letter Capitalized, no trailing period).  The "what and why" is at the end of the list, yet is critical, while some of the formatting desires, while justifiable, are notably less critical, imho.

File: contributors/guide/pull-requests.md (hunk 4 above)
Comment (tpepper) [reply contributors/guide/pull-requests.md:326]: Also, in keeping with the theme of GitHub PR artifacts are not in the git commit history:  We have useful area labels for organizing on GitHub.  Recommending using those, or similar descriptive lead in words, in the summary line would further give that useful organizing to those doing project/code maintenance relative to what you see in the git history.

File: contributors/guide/pull-requests.md (hunk 5 above)
Comment (mrbobbytables) [reply contributors/guide/pull-requests.md:362]: My thought on the ordering was a sort of what you encounter first approach hitting the relevant sections for the subject, how to add the body, what things if found will prevent it from merging. Maybe re-order with the what/why at the top?

File: contributors/guide/pull-requests.md (hunk 3 above)
Comment (mrbobbytables) [reply contributors/guide/pull-requests.md]: Updated headings so auto-toc doesn't include them 👍

File: contributors/guide/pull-requests.md (hunk 5 above)
Comment (mrbobbytables) [reply contributors/guide/pull-requests.md:355]: I left it as is, but expanded a little further on it in the subject line topic and referenced that section.

File: contributors/guide/pull-requests.md (hunk 4 above)
Comment (mrbobbytables) [reply contributors/guide/pull-requests.md]: Referenced our kind and area labels (with links) for commit message prefixes.

File: contributors/guide/pull-requests.md (hunk 5 above)
Comment (tpepper) [reply contributors/guide/pull-requests.md:355]: The latest commit really helps!  Thanks!

**Comment by mrbobbytables** [comment]
I'll let tide squash it all up 👍 
/hold cancel

**Comment by saschagrunert** [comment]
Awesome work, I’m glad to see this got merged.

**Pull request merged**